### PR TITLE
Sync logging out between tabs

### PR DESCRIPTION
**Pull request by PiotrKozlowski**
for: https://github.com/openstax/unified/issues/782

TODO:
- [x] Fix tests

### CODE REVIEW

File: src/gateways/createHighlightClient.ts [hunk 1]
```diff
@@ -7,6 +9,8 @@ const formatError = (response: Response) => {
     });
 };
 
+export class DoNotHandleMe extends Error {}
```
Comment (TomWoodward) [inline src/gateways/createHighlightClient.ts]: please rename this `UnauthenticatedError`, move it to a more generic location, and move the `dispatch(receiveLoggedOut())` to the middleware in app/utils

File: src/app/utils.ts [hunk 2]
```diff
@@ -27,6 +28,9 @@ export const actionHook = <C extends AnyActionCreator>(actionCreator: C, body: A
 
       if (matches(action)) {
         const catchError = (e: Error) => {
```
Comment (TomWoodward) [inline src/app/utils.ts]: could be worth extracting this `catchError` function just to reduce complexity in this function

**Review by TomWoodward — CHANGES_REQUESTED** [review]

File: src/app/types.ts [hunk 3]
```diff
@@ -70,3 +70,5 @@ export type ActionHookBody<C extends AnyActionCreator> = (helpers: MiddlewareAPI
 // helpers
 export type ArgumentTypes<F> = F extends (...args: infer A) => any ? A : never;
 export type FirstArgumentType<F> = F extends (first: infer A, ...args: any) => any ? A : never;
+
+export class UnauthenticatedError extends Error {}
```
Comment (TomWoodward) [inline src/app/types.ts]: this isn't a typescript type, it shouldn't be in here

File: src/app/utils.ts [hunk 4]
```diff
@@ -26,23 +28,28 @@ export const actionHook = <C extends AnyActionCreator>(actionCreator: C, body: A
...
       if (matches(action)) {
-        const catchError = (e: Error) => {
-          Sentry.captureException(e);
-          stateHelpers.dispatch(recordError(e));
-        };
         try {
           const promise = boundHook(action);
           if (promise) {
-            services.promiseCollector.add(promise.catch(catchError));
+            services.promiseCollector.add(promise.catch((e) => catchError(e, stateHelpers.dispatch)));
```
Comment (TomWoodward) [inline src/app/utils.ts]: you could curry this ahead of time so you don't have to pass it in two places

```
const catchError = makeCatchError(stateHelpers);
```

```suggestion
            services.promiseCollector.add(promise.catch(catchError));
```

**Review by TomWoodward — CHANGES_REQUESTED** [review]
in the past there have been problems extending from the base Error class (https://stackoverflow.com/a/35858868) please make sure that this is not still a problem

File: src/gateways/createHighlightClient.spec.ts [hunk 5]
```diff
@@ -1,7 +1,12 @@
 import { GetHighlightsSourceTypeEnum } from '@openstax/highlighter/dist/api';
+import { UnauthenticatedError } from '../app/types';
 import { resetModules } from '../test/utils';
 import createHighlightClient from './createHighlightClient';
 
+jest.mock('../app/utils', () => ({
+  ...jest.requireActual('../app/utils'),
+}));
+
```
Comment (janKozie1) [inline src/gateways/createHighlightClient.spec.ts]: Is this necessary? Won't `jest.mock('../app/utils')` work?

Comment (TomWoodward) [reply src/gateways/createHighlightClient.spec.ts]: actually this is mocking it with the actual module, so wouldn't just not mocking it do the same?

**Comment by PiotrKozlowski** [comment]
@TomWoodward 
> In the past there have been problems extending from the base Error class (https://stackoverflow.com/a/35858868) please make sure that this is not still a problem

We also had this problem with `adaptarr` but it doesn't seem to be a problem when anymore. Probably because we are targeting `"target": "es2017"` and babel doesn't have to compile this code in any way?

File: src/app/utils.ts [hunk 6]
```diff
@@ -26,10 +27,7 @@ export const actionHook = <C extends AnyActionCreator>(actionCreator: C, body: A
       const result = next(action);
 
       if (matches(action)) {
-        const catchError = (e: Error) => {
-          Sentry.captureException(e);
-          stateHelpers.dispatch(recordError(e));
-        };
+        const catchError = makeCatchError(stateHelpers.dispatch);
```
Comment (TomWoodward) [inline src/app/utils.ts]: can you hoist this up to line 22ish?

**Review by TomWoodward — CHANGES_REQUESTED** [review]
i left one little comment but i'm sending this to qa

**Review by TomWoodward — APPROVED** [review]

**Pull request merged**